### PR TITLE
Improve `arange`

### PR DESCRIPTION
**Pull request by Theodor-Lindberg**
# PR Summary

The first implementation of `arange` generated the values using Python floats and then casted the list to the destination format. 
However Python floats cannot represented all the values APyTypes can which creates some numerical problems.

The new solution generates the values in the C++ layer using APyFixed, to then cast the values to the destination format.

# PR Checklist

- [N/A] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is tested
- [x] relevant changes added to [CHANGELOG.md](https://github.com/apytypes/apytypes/blob/main/CHANGELOG.md)
- [x] new functionality is documented

### CODE REVIEW

File: src/apyfixed.cc [hunk 1]
```diff
@@ -995,6 +995,45 @@ APyFixed APyFixed::from_double(
...
+APyFixed APyFixed::from_unspecified_double(double value)
+{
+    if (std::isnan(value) || std::isinf(value)) {
+        throw std::domain_error(fmt::format("Cannot convert {} to fixed-point", value));
+    }
+
+    int exp = exp_of_double(value);
+    uint64_t man = man_of_double(value);
+
+    // Append mantissa hidden one
```
Comment (miklhh) [inline src/apyfixed.cc:1007]: Look at the new `APyFixed.set_from_double()` for properly handling subnormals.

Or better still, let this function only evaluate the necessary width of the resulting APyFixed, and reuse `APyFixed.set_from_double()` for the data casting.

File: src/apyfixed.cc [hunk 2]
```diff
@@ -995,6 +995,45 @@ APyFixed APyFixed::from_double(
     return result;
 }
 
+APyFixed APyFixed::from_unspecified_double(double value)
```
Comment (miklhh) [inline src/apyfixed.cc:998]: Look at the new `APyFixed.set_from_double()` for properly handling subnormals.

Or better still, let this function only evaluate the necessary width of the resulting APyFixed, and reuse `APyFixed.set_from_double()` for the data casting.

Comment (Theodor-Lindberg) [reply src/apyfixed.cc:995]: Thanks, fixed.
Re-using the function `set_from_double` is certainly good for maintenance, but the function is optimized to generate small word lengths that can accommodate the data while making use of the fact the result is always one or two limbs.

File: src/array_utils.h [hunk 3]
```diff
@@ -151,4 +153,61 @@ static APY_INLINE void transpose_axes_and_copy_data(
...
+    if (nb::isinstance<APyFloat>(val)) {
+        const auto v = static_cast<APyFloat>(nb::cast<APyFloat>(val));
+        return v.to_fixed();
+    } else if (nb::isinstance<nb::int_>(val)) {
+        return APyFixed::from_unspecified_integer(nb::cast<nb::int_>(val));
+    } else if (nb::isinstance<nb::float_>(val)) {
+        const auto v = static_cast<double>(nb::cast<nb::float_>(val));
+        const auto fx = APyFixed::from_unspecified_double(v);
+        return fx;
+    } else { // APyFixed
```
[the remark this answers is not in the document]
Comment (Theodor-Lindberg) [reply src/array_utils.h]: Fixed.

File: src/apyfixedarray_wrapper.cc [hunk 4]
```diff
@@ -1231,7 +1231,38 @@ void bind_fixed_array(nb::module_& m)
...
+            "arange",
+            &APyFixedArray::arange,
+            nb::arg("start"),
+            nb::arg("stop"),
+            nb::arg("step"),
+            nb::arg("int_bits") = nb::none(),
+            nb::arg("frac_bits") = nb::none(),
+            nb::arg("bits") = nb::none(),
+            R"pbdoc(
+            Initializes an array filled with the specified value.
```
Comment (miklhh) [inline src/apyfixedarray_wrapper.cc:1244]: * `arange(stop)`: Values are generated within the half-open interval [0, stop) (in other words, the interval including start but excluding stop).
*  `arange(start, stop)`: Values are generated within the half-open interval [start, stop).
*  `arange(start, stop, step)`: Values are generated within the half-open interval [start, stop), with spacing between values given by step.

File: lib/apytypes/_array_functions.py [hunk 5]
```diff
@@ -851,19 +851,20 @@ def arange(
     frac_bits=None,
     bits=None,
     exp_bits=None,
-    mantissa_bits=None,
+    man_bits=None,
+    bias=None,
 ):
     """
-    Simple and naive version of arange.
+    Create an array with evenly spaced values within a given interval.
```
Comment (miklhh) [inline lib/apytypes/_array_functions.py:858]: * `arange(stop)`: Values are generated within the half-open interval [0, stop) (in other words, the interval including start but excluding stop).
*  `arange(start, stop)`: Values are generated within the half-open interval [start, stop).
*  `arange(start, stop, step)`: Values are generated within the half-open interval [start, stop), with spacing between values given by step.

File: src/apyfloatarray_wrapper.cc [hunk 6]
```diff
@@ -652,6 +652,38 @@ void bind_float_array(nb::module_& m)
...
+            "arange",
+            &APyFloatArray::arange,
+            nb::arg("start"),
+            nb::arg("stop"),
+            nb::arg("step"),
+            nb::arg("exp_bits"),
+            nb::arg("man_bits"),
+            nb::arg("bias") = std::nullopt,
+            R"pbdoc(
+            Initializes an array filled with the specified value.
```
Comment (miklhh) [inline src/apyfloatarray_wrapper.cc:666]: * `arange(stop)`: Values are generated within the half-open interval [0, stop) (in other words, the interval including start but excluding stop).
*  `arange(start, stop)`: Values are generated within the half-open interval [start, stop).
*  `arange(start, stop, step)`: Values are generated within the half-open interval [start, stop), with spacing between values given by step.

**Pull request merged**